File: python/023_merge_k_sorted_lists.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        head = pointer = ListNode(0)
        node_val_list = []
        for l in lists:
            while l:
                node_val_list.append(l.val)
                l = l.next

        node_val_list.sort()

        for i in node_val_list:
            pointer.next = ListNode(i)
            pointer = pointer.next

        return head.next


if __name__ == "__main__":
    s = Solution()

    h1 = ListNode(1)
    h2 = ListNode(2)
    h3 = ListNode(4)
    h1.next = h2
    h2.next = h3

    l1 = ListNode(1)
    l2 = ListNode(3)
    l3 = ListNode(4)

    l1.next = l2
    l2.next = l3

    print(s.mergeKLists([h1, l1]))

File: cpp/linked_list/023_merge_k_sorted_lists.cpp
#include "../header.h"
using namespace std;

struct ListNode {
    int val;
    ListNode *next;  // 一个指针存下个ListNode
    ListNode(int x) : val(x), next(NULL) {}
 };

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 1) return lists[0];
        ListNode dummy(0);
        ListNode *tail = &dummy;  // 对指针进行操作而不是对对象进行操作
        vector<int> q;
        int n = lists.size();
        
        for (int i = 0; i < n; i++) {
            ListNode *l = lists[i];
            while (l) {
                int v = l->val;  // 对指向struct的指针，取地址里面的值，用->
                if (l) q.push_back(v);
                l = l->next;
            }        
        }
        sort(q.begin(),q.end());  // nlog_2(n)
        int m = q.size();
        for (int i = 0; i < m; i++) {
            // new： 为变量请求新内存， 返回地址
            ListNode *temp = new ListNode(q[i]);
            tail->next = temp;
            tail = tail->next;  // 对指向struct的指针，取地址里面的值，用->
        }
        return dummy.next;
    }
};

int main() {
    Solution s;
    vector<ListNode*> lists;
    ListNode l1(1);
    ListNode l2(1);
    ListNode l3(2);
    
    ListNode temp(4);
    l1.next = &temp;
    ListNode temp2(5);
    temp.next = &temp2;

    ListNode temp3(3);
    l2.next = &temp3;
    ListNode temp4(4);
    temp.next = &temp4;

    ListNode temp5(6);
    l3.next = &temp5;

    lists.push_back(&l1);
    lists.push_back(&l2);
    lists.push_back(&l3);

    cout << s.mergeKLists(lists)->val << endl;
}